File: Dynamic Programming/Python/Minimum Number of Removals to Make Mountain Array.py
# Problem link: https://leetcode.com/problems/minimum-number-of-removals-to-make-mountain-array/description
# Time: O(n ^ 2)
# Space: O(n)
class Solution:
    def minimumMountainRemovals(self, nums: List[int]) -> int:
        size = result = len(nums)
        LIS, LDS = [1] * size, [1] * size

        for pntOne in range(size):
            for pntTwo in range(pntOne):
                if nums[pntTwo] < nums[pntOne]:
                    LIS[pntOne] = max(LIS[pntOne], 1 + LIS[pntTwo])
        
        for pntOne in range(size - 1, -1, -1):
            for pntTwo in range(size - 1, pntOne, -1):
                if nums[pntTwo] < nums[pntOne]:
                    LDS[pntOne] = max(LDS[pntOne], 1 + LDS[pntTwo])

        for index in range(1, size - 1):
            if LIS[index] > 1 and LDS[index] > 1:
                result = min(size - LIS[index] - LDS[index] + 1, result)
        
        return result

File: Dynamic Programming/Java/Minimum Number of Removals to Make Mountain Array.java
// Problem link: https://leetcode.com/problems/minimum-number-of-removals-to-make-mountain-array/description
// Time: O(n ^ 2)
// Space: O(n)
class Solution {
    public int minimumMountainRemovals(int[] nums) {
        int result = nums.length, size = nums.length;
        int[] LIS = new int[size], LDS = new int[size];
        Arrays.fill(LIS, 1);
        Arrays.fill(LDS, 1);

        for(int pntOne = 0; pntOne < size; pntOne++) {
            for(int pntTwo = 0; pntTwo < pntOne; pntTwo++) {
                if(nums[pntTwo] < nums[pntOne]) {
                    LIS[pntOne] = Math.max(LIS[pntOne], 1 + LIS[pntTwo]);
                }
            }
        }

        for(int pntOne = size - 1; pntOne >= 0; pntOne--) {
            for(int pntTwo = size - 1; pntTwo > pntOne; pntTwo--) {
                if(nums[pntTwo] < nums[pntOne]) {
                    LDS[pntOne] = Math.max(LDS[pntOne], 1 + LDS[pntTwo]);
                }
            }
        }

        for(int index = 1; index < size - 1; index++) {
            if(LIS[index] > 1 && LDS[index] > 1) {
                result = Math.min(result, size - LIS[index] - LDS[index] + 1);
            }
        }

        return result;
    }
}
